### PR TITLE
Fix makepipelineplan

### PR DESCRIPTION
**Pull request by warren830**
# Summary

modify github blueprint.go and blueprint_test.go to be more elegant

### Does this close any open issues?
relate to #3127 

### Screenshots
image

### Other Information
Any other information that is important to this PR.

### CODE REVIEW

File: plugins/github/api/blueprint.go [hunk 1]
```diff
@@ -38,179 +38,178 @@ import (
...
+			}
+			if err != nil {
+				return nil, err
+			}
+			repo, err = getApiRepoIfNil(op)
 			if err != nil {
 				return nil, err
 			}
+
+			doraOption := make(map[string]interface{})
```
Comment (klesh) [inline plugins/github/api/blueprint.go:173]: Can we move 173 ~ 177 to 182 as part of the stage in initialize style declaration to make it more readable?

File: plugins/github/api/blueprint.go [hunk 2]
```diff
@@ -38,179 +38,178 @@ import (
...
-	stage := plan[i]
-	if stage == nil {
-		stage = core.PipelineStage{}
+	return plan, nil
+}
+
+func makePipelinePlan(subtaskMetas []core.SubTaskMeta, scope []*core.BlueprintScopeV100, apiClient helper.ApiClientGetter, connection *models.GithubConnection) (core.PipelinePlan, errors.Error) {
+	var err errors.Error
+	var repo *tasks.GithubApiRepo
+	getApiRepoIfNil := func(op *tasks.GithubOptions) (*tasks.GithubApiRepo, errors.Error) {
```
Comment (klesh) [inline plugins/github/api/blueprint.go:71]: How about `memorizedGetApiRepo`?

Comment (klesh) [reply plugins/github/api/blueprint.go:71]: This memorized function should be moved after line 112 and take no parameters.
It makes no sense if the caller passed a different `op`

File: plugins/github/api/blueprint.go (hunk 1 above)
Comment (warren830) [reply plugins/github/api/blueprint.go]: fixed

File: plugins/github/api/blueprint.go (hunk 2 above)
Comment (warren830) [reply plugins/github/api/blueprint.go]: fixed

File: plugins/github/api/blueprint.go [hunk 3]
```diff
@@ -38,179 +38,180 @@ import (
...
+		options["connectionId"] = connection.ID
+		options["transformationRules"] = transformationRules
+		// make sure task options is valid
+		op, err := tasks.DecodeAndValidateTaskOptions(options)
+		if err != nil {
+			return nil, err
 		}
-		if j == len(plan) {
-			plan = append(plan, nil)
+		memorizedGetApiRepo := func(op *tasks.GithubOptions) (*tasks.GithubApiRepo, errors.Error) {
```
Comment (klesh) [inline plugins/github/api/blueprint.go:110]: Please remove `op *tasks.GithubOptions`, let function uses the `op` variable from 106 instead.

File: plugins/github/api/blueprint_test.go [hunk 4]
```diff
@@ -66,19 +81,58 @@ func TestProcessScope(t *testing.T) {
...
 	scopes := make([]*core.BlueprintScopeV100, 0)
 	scopes = append(scopes, bs)
-	plan := make(core.PipelinePlan, len(scopes))
-	for i, scopeElem := range scopes {
-		plan, err = processScope(nil, 1, scopeElem, i, plan, apiRepo, connection)
-		assert.Nil(t, err)
+	plan, err := makePipelinePlan(nil, scopes, mockApiCLient, connection)
+	assert.Nil(t, err)
+
+	//planJson, err1 := json.Marshal(plan)
```
Comment (klesh) [inline plugins/github/api/blueprint_test.go:89]: Plese remove commented code

**Review by abeizn — APPROVED** [review]
LGTM

**Pull request merged**